File: Source/Aladdin.CAPI.Environment/ConfigRandFactory.cs
using System;
using System.Diagnostics.CodeAnalysis; 

namespace Aladdin.CAPI
{
    ///////////////////////////////////////////////////////////////////////////
    // Фабрика создания генераторов случайных данных
    ///////////////////////////////////////////////////////////////////////////
    [SuppressMessage("Microsoft.Design", "CA1063:ImplementIDisposableCorrectly")]
    public class ConfigRandFactory : RefObject, IRandFactory
    {
        // фабрика создания генераторов случайных данных
        private IRandFactory randFactory; private bool critical; 

        // конструктор
        public ConfigRandFactory(IRandFactory randFactory, bool critical)
        {
            // сохранить переданные параметры
            this.randFactory = RefObject.AddRef(randFactory); this.critical = critical; 
        }
        // освободить используемые ресурсы
        protected override void OnDispose()
        {
            // освободить используемые ресурсы
            RefObject.Release(randFactory); base.OnDispose(); 
        }
        // создать генератор случайных данных
        public virtual IRand CreateRand(object window)
        {
            // создать генератор случайных данных
            try { return randFactory.CreateRand(window); }

            // обработать возможную ошибку
            catch { if (critical) throw; return null; }
        }
    }
}

File: Source/Aladdin.CAPI.Environment.Java/src/aladdin/capi/ConfigRandFactory.java
package aladdin.capi;
import aladdin.*; 
import java.io.*; 

///////////////////////////////////////////////////////////////////////////
// Фабрика создания генераторов случайных данных
///////////////////////////////////////////////////////////////////////////
public class ConfigRandFactory extends RefObject implements IRandFactory
{
    // фабрика создания генераторов случайных данных
    private final IRandFactory randFactory; private final boolean critical; 

    // конструктор
    public ConfigRandFactory(IRandFactory randFactory, boolean critical)
    {
        // сохранить переданные параметры
        this.randFactory = RefObject.addRef(randFactory); this.critical = critical; 
    }
    // освободить используемые ресурсы
    @Override protected void onClose() throws IOException  
    {
        // освободить используемые ресурсы
        RefObject.release(randFactory); super.onClose(); 
    }
    // создать генератор случайных данных
    @Override public IRand createRand(Object window) throws IOException
    {
        // создать генератор случайных данных
        try { return randFactory.createRand(window); }

        // обработать возможную ошибку
        catch (Throwable e) { if (critical) throw e; return null; }
    }
}
